### ISSUE TITLE
Confusing Syntax for Error Handling

### ISSUE THREAD

**Issue by Dr4gon** [franca/franca#201]
Hi,

I tried to use franca IDL to design a general error enum in Java for several interface methods. Below you'll find the snippet without type collection and so on.

```
enumeration ErrorEnum {
     /**
        * Bad request, illegal parameter set (e.g. parameter missing).
        * The client shall not repeat this request without modification.
        */
       BAD
       /**
        * Not found, the requested data was not found.
        * The client shall not repeat this request without modification.
        */
       NOT_FOUND
       /**
        * Internal error during request processing.
        * The client may retry the request later.
        */
       INTERNAL_ERROR
       /**
        * During request processing a call to an external system was started but failed.
        * The client may retry the request later but there is no guaranty of successful execution.
        */
       GATEWAY_ERROR
       /**
        * During request processing a call to an external system was started but timed out.
        * The client may retry the request later.
        */
       GATEWAY_TIMEOUT
  } 

error{
   MyPackageName.ErrorEnum error
}
```

However I figured out that this is not possible because the syntax should be:

```
error MyPackageName.ErrorEnum error
```

I think that's rather confusing to the user because a custom out looks like this:

```
out{
 MyPackageName.CustomStruct return
}
```

As a proposal: Maybe you can consider changing that at least to me odd syntax behaviour in a future version?

Best regards
Dr4gon

**Comment by kbirken** [comment]
I am not sure about what part of the syntax seems odd to you... esp. because your examples with "error" are out of context. Here is a proper example i the context of a Franca method (which is the only context where errors might be used):

```
method divide {
  in { ... }
  out { ... }

  error {
    DIVISION_BY_ZERO
    OVERFLOW
    UNDERFLOW
  }
}
```

This example is an inline error definition, which is kind of an ad-hoc enumeration. Your use case (referring an external enumeration type) might look like this:

```
method divide {
  in { ... }
  out { ... }
  error ErrorEnum
}
```

There is a third option, which is to mix both above options (extend a referenced external enumeration)

```
method divide {
  in { ... }
  out { ... }
  error extends ErrorEnum {
    ANOTHER_ERROR
    YET_ANOTHER_ERROR
  }
}
```

For which of these options do you like to propose syntax changes, and which ones?

**Comment by Dr4gon** [comment]
Here's the more detailed explanation refering to your example. 

I'm talking about the second option:
```
method divide {
  in { ... }
  out { ... }
  error ErrorEnum
}
```
I'd expected it to look like this:
```
method divide {
  in { ... }
  out { ... }
  error {
     ErrorEnum error
   }
}
```
Because a custom struct in an out is returned like this:
```
method divide {
  in { ... }
  out { 
     MyOwnStruct result
   }
  error {
     ErrorEnum error
   }
}
```

You might point me to the first option for my use case here. That only works if I duplicate the enum for every method which has the same errors. Which I don't want to do because changing it means maintaining all the definitions.

The third option in my understanding isn't viable either because it extends the error. I want to generalize it not extending it.

This is what I want to do:

```
method divide {
  in { ... }
  out { ... }
  error {
     ErrorEnum error
   }
}

method add {
  in { ... }
  out { ... }
  error {
     ErrorEnum error
   }
}

method substract {
  in { ... }
  out { ... }
  error {
     ErrorEnum error
   }
}
```

Does this answer your question?

**Comment by kbirken** [comment]
I agree that the second option (see my comment on 2016-12-04) is not consistent with the syntax of the `out { ... }` section of a method. However, the in- and out-sections specify method parameters and return values, the error-section specifies error codes which could be returned by the method. We designed the language syntax such that all three options can be used without much syntactic overhead (i.e., defining some error enumerators ad-hoc, referring to an existing enum type and extending an existing enum type). It is not clear to me how the first and third option would look like if the syntax of error- and out-section are aligned.
